### ISSUE TITLE
Initial value of disabled attribute is not rendered as expected

### ISSUE THREAD

**Issue by flosse** [seed-rs/seed#261]
```rust
button![attrs{At::Disable => false;}]
```
renders to
```html
<button disabled="false"></button>
```
instead of
```html
<button></button>
```

See #260

**Comment by MartinKavik** [comment]
It's not bug but feature, see https://github.com/David-OConnor/seed/blob/afd4fa8387f40e2aaf7f07cb41a36da68f0b808e/src/dom_types/values.rs#L44

It was pain to maintain hardcoded DB of attributes which should or don't should be hidden if you set it's value to something falsy. And there were often bugs if I remember correctly.

We want to have auto-generated types for Tags, Attributes, Events and Styles. Then we want to type somehow also Attribute values and Style values - it should prevent from surprising behavior like you described. 
Styles are the only entity now, that is populated from DB to generate Rust types - DB is here: https://github.com/MartinKavik/html-css-db.

**Comment by flosse** [comment]
Yea, typed values are great, but why is there still a `"false"` string when the should be nothing?

**Comment by MuhannadAlrusayni** [comment]
I think the conversion should happen inside the attrs! macro some how.

**Comment by MartinKavik** [comment]
@flosse:

> why is there still a "false" string when there should be nothing?

You need to call method `.as_at_value()`, e.g. `attrs!{ At::Disabled => false.as_at_value() }`. Everything is converted to `String` and taken as the valid value by default - https://github.com/David-OConnor/seed/blob/afd4fa8387f40e2aaf7f07cb41a36da68f0b808e/src/dom_types/values.rs#L59
We can't implement `ToString` for `AtValue` and `bool` for `AtValue` at the same time on the `stable` Rust, so calling `as_at_value` is a trade-off. Otherwise you would have to implement only specific types for `AtValue`.
Another reason for trait `AsAtValue` - 
https://github.com/David-OConnor/seed/blob/afd4fa8387f40e2aaf7f07cb41a36da68f0b808e/src/dom_types/values.rs#L75
\- you can implement it for your types and use - e.g. `Option<X>` instead of  `bool` - as the attribute value.

So yeah, there is space for improvements - you can try design better API or we can wait until the `specialization` is integrated into `stable` Rust.

---

Another problem with bool as attribute value - what false or true really mean? E.g.:
- attribute `disabled`
  - `false` => hide the entire attribute
  - `true` =>  `<button disabled>` or `<button disabled="disabled">` or other string?

- attribute `autocomplete`
  - `false` => hide attribute or `autocomplete="off"`?
  - `true` => `autocomplete` or `autocomplete="on"` or other string?

- and what about, for example, three-state custom attributes with possible values `"false"`, `"true"` and _not set_?

So I think everyone can have different opinion about false and true.

---

I was also experimenting with other optons - https://github.com/David-OConnor/seed/issues/182#issuecomment-509971846 - but the current API seems to be the best trade-off so far from my point of view.

**Comment by MartinKavik** [comment]
@MuhannadAlrusayni 
You can write trait `AsCSSValue` (similar to `AsAtValue`) and combine it with typed values:

<details>
<summary>Code</summary>

```rust
enum AlignItems {
    Stretch,
    Center,
    Start,
    End,
}

// or use https://github.com/Peternator7/strum
impl fmt::Display for AlignItems {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let value = match self {
            Self::Stretch => "stretch",
            Self::Center => "center",
            Self::Start => "start",
            Self::End => "end",
        };
        write!(f, "{}", value)
    }
}

pub trait AsCSSValue {
    fn as_css_value(&self) -> CSSValue;
}

impl AsCSSValue for Option<AlignItems> {
    fn as_css_value(&self) -> CSSValue {
        self
            .as_ref()
            .map(ToString::to_string)
            .map(CSSValue::Some)
            .unwrap_or(CSSValue::Ignored)
    }
}

fn view(model: &Model) -> impl View<Msg> {
    let align_items: Option<AlignItems> = Some(AlignItems::Stretch);

    div![
        style![
            St::AlignItems => align_items.as_css_value(),
``` 
</details>

Or you can try to design new API - e.g. macro calls `.as_css_value` implicitly or/and converts `Option` to `CSSValue` automatically...
Or start to design typed CSS for Seed.
I didn't have time for it, so when the time comes we will have some foundations to build on.

**Comment by MuhannadAlrusayni** [comment]
> Or you can try to design new API - e.g. macro calls .as_css_value implicitly.
Yes that what I meant earlier.

I think we should have both a trait ToCssValue and use that trait in the `style!` macro to convert values to CSSValue..

should I open PR and try to do that or there are other planes for this issue ?

**Comment by MartinKavik** [comment]
@MuhannadAlrusayni 
> should I open PR and try to do that or there are other planes for this issue ?

Yes, please.
`ToCssValue` vs `AsCssValue` - I chose `as_` because in the most cases the conversion is almost free, but it depends ([naming docs](https://rust-lang-nursery.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)).

**Comment by flosse** [comment]
> > why is there still a "false" string when there should be nothing? 
>>
> You need to call method `.as_at_value()`

I'm sorry, I wasn't aware of this.

> Another problem with bool as attribute value - what false or true really mean?

good point!
 
> the current API seems to be the best trade-off so far from my point of view.

yes, for now the current API is working for me (now that I know how to do it right ;-))
In the future I think we can do better (whatever it looks like).

**Comment by MartinKavik** [comment]
> I'm sorry, I wasn't aware of this.

I think it isn't in docs. Once the docs design with simple search is complete I want to update content in cooperation with @David-OConnor.

> In the future I think we can do better (whatever it looks like).

Let's wait for @MuhannadAlrusayni's PR (see previous comments). Once updated API for CSS values are designed/implemented, we can focus on attribute values.

**Comment by David-OConnor** [comment]
I've updated the structure section of the homepage to show use of `.as_at_value()`.

**Comment by MartinKavik** [comment]
> Let's wait for @MuhannadAlrusayni's PR (see previous comments). Once updated API for CSS values are designed/implemented, we can focus on attribute values.

- style! PR: https://github.com/David-OConnor/seed/pull/273/files
- test with `false` as the atttibute: https://github.com/David-OConnor/seed/pull/260/files

**Comment by MartinKavik** [comment]
Closing in favor of https://github.com/seed-rs/seed/issues/312
